### PR TITLE
Add and Remove ZeRO 3 Hooks

### PR DESCRIPTION
**Pull request by jomayeri**
Gives the ability to add and remove the forward hooks in ZeRO 3 by using a context manager. These code changes were taken from a Huggingface [PR](https://github.com/huggingface/trl/pull/1617) and integrated for direct support in DeepSpeed.

This is useful in the inference case and the speedup can be observed [here](https://github.com/huggingface/trl/pull/1483).

### CODE REVIEW

**Comment by tjruwase** [comment]
@jomayeri, please add some unit tests

File: deepspeed/runtime/zero/utils.py [hunk 1]
```diff
@@ -158,3 +159,36 @@ def apply_to_tensors_only(function, value, warning_msg_fn=None):
                 logger.warning(warning_msg_fn(value))
                 warned = True
         return value
+
+
+@contextmanager
+def unwrap_model_for_generation(model):
```
Comment (tjruwase) [inline deepspeed/runtime/zero/utils.py:165]: It is better to use general naming that describes what the utility does rather than specific usage like generation. 

```suggestion
def unshard_and_remove_hooks(model):
```

File: deepspeed/runtime/zero/utils.py [hunk 2]
```diff
@@ -158,3 +159,36 @@ def apply_to_tensors_only(function, value, warning_msg_fn=None):
...
+    For ZeRO-3 models, we gather the weights once to speed up generation.
+    """
+    with deepspeed.zero.GatheredParameters(model.parameters()):
+        # Removes the optimizer hooks from a DeepSpeed ZeRO-3 model.
+
+        # Remove hooks
+        if model.optimizer is not None and hasattr(model.optimizer, "parameter_offload"):
+            optimizer_offload = model.optimizer.parameter_offload
+        elif model.optimizer is not None:
+            optimizer_offload = model.optimizer
```
Comment (tjruwase) [inline deepspeed/runtime/zero/utils.py:176]: Can you explain this zero-3 case where we have hooks attached to the optimizer?

File: csrc/aio/py_lib/deepspeed_py_copy.cpp [hunk 3]
```diff
@@ -10,7 +10,7 @@ Functionality for swapping optimizer tensors to/from (NVMe) storage devices.
 #include "deepspeed_py_copy.h"
 #include <omp.h>
 
-#define ROUND_DOWN(size, step) ((size) & ~((step)-1))
+#define ROUND_DOWN(size, step) ((size) & ~((step) - 1))
```
Comment (tjruwase) [inline csrc/aio/py_lib/deepspeed_py_copy.cpp:13]: Is this related?

File: csrc/deepspeed4science/evoformer_attn/iterators/predicated_tile_access_iterator_residual_last.h [hunk 4]
```diff
@@ -488,7 +488,7 @@ class PredicatedTileAccessIteratorResidualLast<Shape_,
 
         /// Construct the Params object given a pitch-linear tensor's layout
         CUTLASS_HOST_DEVICE
-        Params(Layout const& layout) : params_(layout::PitchLinear(layout.stride(0))){};
+        Params(Layout const& layout) : params_(layout::PitchLinear(layout.stride(0))) {};
```
Comment (tjruwase) [inline csrc/deepspeed4science/evoformer_attn/iterators/predicated_tile_access_iterator_residual_last.h:491]: Is this related?

File: tests/unit/runtime/zero/test_zero_context.py [hunk 5]
```diff
@@ -299,3 +300,6 @@ def test(self):
         with deepspeed.zero.GatheredParameters(l.weight):
             # all ranks compare
             assert torch.equal(l.weight, torch.zeros_like(l.weight))
+
```
Comment (tjruwase) [inline tests/unit/runtime/zero/test_zero_context.py:303]: Is this needed?

File: csrc/aio/py_lib/deepspeed_py_copy.cpp (hunk 3 above)
Comment (loadams) [reply csrc/aio/py_lib/deepspeed_py_copy.cpp:13]: I think this is due to formatting, we probably should update the clang-format version we use too.

File: csrc/deepspeed4science/evoformer_attn/iterators/predicated_tile_access_iterator_residual_last.h (hunk 4 above)
Comment (loadams) [reply csrc/deepspeed4science/evoformer_attn/iterators/predicated_tile_access_iterator_residual_last.h]: Fixed with #5839 as well.

**Review by tjruwase — APPROVED** [review]

**Comment by tjruwase** [comment]
@jomayeri, can you please add documentation of this new API somewhere below?

https://deepspeed.readthedocs.io/en/latest/zero3.html
